Add master DRBG for reseeding

mspncp opened the pull request:
##### Checklist

- [x] documentation is added or updated  (as comments)
- [x] tests are added or updated

# The master DRBG

Up to now, there was an asymmetry between the public and the private global DRBG. Only the former was reseeded by the `RAND_add()` call. This pull request introduces a master DRBG which is not publicly accessible. So there are now three global DRBGs:

```
 <os entropy sources>
        |
     <master>          \
      /   \            | three global DRBGs with locking
<public>  <private>    /
    |
  <ssl>
```
The master DRBG is the only DRBG that pulls entropy from the trusted entropy sources, and that is reseeded by the `RAND_add()` call. Three of the DRBGs, namely master, public and private DRBG, have a locking mechanism, while the DRBGs chained to the public DRBG are intended to be used by a single thread and have no locking.

## Naming

In order to get a consistent naming of the three global DRBGs, the two existing DRBGs had to be renamed. This includes the `RAND_DRBG_get0_*()` method which are exported from libcrypto to libssl, but that's no problem because these functions were introduced in master only recently and are not part of an official API yet.

### Module globals

```C
static RAND_DRBG drbg_master;  /* The master DRBG. Seeded from the os */
static RAND_DRBG drbg_public;  /* The global public DRBG. */
static RAND_DRBG drbg_private; /* The global private-key DRBG. */
``` 

### Accessors

```C
RAND_DRBG *RAND_DRBG_get0_master(void);
RAND_DRBG *RAND_DRBG_get0_public(void);
RAND_DRBG *RAND_DRBG_get0_private(void);
```

## Seed Propagation

Instead of having `RAND_add()` reseed all global DRBGs (which would require taking all three locks), we only reseed the master DRBG (because it holds the lock already) and use a simple mechanism to propagate the reseeding to to all chained DRBGs: Every DRBG has a seed count which is incremented whenever the DRBG pulls fresh random input. Any chained DRBG compares in the `RAND_DRBG_generate()` call its seed count with the one of its parent. If there is a mismatch, it seeds itself automatically.

This propagation works for regular reseedings, too, not only for the ones triggered by `RAND_add()`.

**Update:** _As suggested by @paulidale, seed propagation will not proceed to the DRBGs which are owned by a thread or an SSL context, because this would make `RAND_add()` a very expensive operation. Only the global public and private DRBGs will be automatically restarted._  

**Remark:** This "seed count" should have been named `seed_count`, however this name unfortunately collides with the `reseed_counter` member which actually counts generate requests, not reseeds. However, the name `reseed_counter` comes from the NIST standard, so I didn't dare to rename it. Hence I called my counter the `restart_count`.  If anybody has a better suggestion for the name, I'm happy to hear about it.

## Locking

Currently, only the three global DRBGs have a lock. The other chained (SSL) locks are considered to be owned by a single thread. So all locking code needs to check for the presence of the lock first. The lock is currently taken by the caller and it is not clear, whether all cases are currently handled. This is somewhat unsatisfying.

* The lock is owned by the RAND_DRBG "class", which should be an opaque structure, in particular the locking should be managed entirely from within the RAND_DRBG "methods". (Sorry for my C++ speak).
* I doubt that the performance gain of having an rwlock only in some dedicated global DRBGs does not justifies destroying simplicity. So why not have an rwlock in all DRBGs and don't have  to check for the presence of the rwlock. (But I'm not an expert w.r.t. the performance of rwlocks.)

## Debugging

You can debug into the code paths of `RAND_add()` using the `openssl rand` command:

```
echo 'this is some random input' > /tmp/randfile
./util/shlib_wrap.sh gdb --args ./apps/openssl rand -hex -rand /tmp/randfile  12

(gdb) start
(gdb) b drbg_setup
(gdb) b drbg_add
```

paulidale commented on crypto/rand/drbg_lib.c:
```diff
@@ -366,6 +376,15 @@ int RAND_DRBG_generate(RAND_DRBG *drbg, unsigned char *out, size_t outlen,
+    if (drbg->parent != NULL) {
+        if (drbg->parent->lock)
+            CRYPTO_THREAD_read_lock(drbg->parent->lock);
+        if (drbg->restart_count != drbg->parent->restart_count)
```
Is the locking needed here?  I thought integer reads were considered atomic.

It might require an atomic increment when the count is changed but that's cheaper than locking when it is available.

mspncp replied:
Then I just leave it as it is: Only the three global DRBGs are created with a lock, the other DRBGs don’t have a lock by default. For completeness sake, I would add a RAND_DRBG_enable_locking() method to the RAND_DRBG API.

I agree with you that letting `CRYPTO_THREAD_xxx_lock()` accept NULL pointers as no-ops is probably not a good idea. So the call needs to be wrapped by a conditional.

mspncp replied:
Thanks for suggesting the atomic operations.  I hope the following question does not sound naive: I was wondering whether any special precaution is necessary in the first place. Even in the unlikely case that due to concurrency issues the if-block

```
if (drbg->restart_count != drbg->parent->restart_count) {
   …
}
```
is not entered erroneously, this would only mean that automatic restarting of the chained DRBG would be postponed until the next generate request. And failing twice in a row is extremely unlikely. What do you think about this kind of reasoning?

mspncp replied:
I would prefer the CRYPTO_atomic_read(), not for performance reasons but for clarity. Provided that's not too much effort for you to add.

paulidale replied:
CRYPTO_atomic_read sounds like a good idea.

paulidale replied:
If there is a race reading either `drbg->restart_count` or `drbg->parent->restart_count`, a reseed will occur now and again the next time the DRBG is used.  I.e. not a big concern.  If a stale value remains in cache, the reseed will be delayed.

paulidale replied:
See #4414

paulidale commented on crypto/rand/rand_lib.c:
```diff
@@ -152,10 +152,16 @@ size_t rand_drbg_get_entropy(RAND_DRBG *drbg,
                                                       bytes_needed,
                                                       8 * bytes_needed);
+                if (drbg->parent->lock)
+                    CRYPTO_THREAD_read_lock(drbg->parent->lock);
```
Unnecessary locking again?

paulidale commented on crypto/rand/drbg_lib.c:
```diff
@@ -366,6 +376,15 @@ int RAND_DRBG_generate(RAND_DRBG *drbg, unsigned char *out, size_t outlen,
+        if (drbg->parent->lock)
+            CRYPTO_THREAD_read_lock(drbg->parent->lock);
+        if (drbg->restart_count != drbg->parent->restart_count)
+            drbg->state = DRBG_RESEED;
```
Certainly the global public and private DRBGs should reseed when the master does.  I'm less sure about the levels below these.

There could be hundreds or thousands of DRBGs in the SSL objects and this would make a reseeding an exceedingly expensive operation.  I recognise that the child DRBGs will reseed in a lazy manner so the cost is spread over time but it's still a large cost.

mspncp replied:
I understand your concerns about the reseeding chain reaction. As you suggested, I could enable automatic restarting only for the  two global chained DRBGs (private and public). There are three possibilities to selectively enable this auto-restart feature for DRBGs:

1.	Use `lock != 0` as an indicator that automatic restart is enabled.  =>  _Looks hacky to me_.
2.	Add an `auto_restart` member.  =>  _Adds a seldomly used member: Most DRBGs will have it disabled,  so there are two unused members, auto_restart and restart_count_.
3.	Use `restart_count != 0` as indicator that automatic restart is enabled. In order to enable auto restart, one simply sets the counter to 1.  An integer overflow check when incrementing can safely be omitted. _Before this ever overflows, madness rules_ ;-).

As usual, the third variant is my favourite choice. :-)  What is your opinion?

paulidale replied:
4. If we aren't reseeding from a parent, reseed the global public and private DRBGs.  It's a big hacky but could be reworked to "if we're reseeding the global master, flag the other two global RNGs for reseeding".

5. Have a global variable indicating that all DRBGs should be reseeded rather than just the global ones.  It wouldn't need locking since it will essentially be read-only.  I can't think of a case where you want some but not all of the non-global DRBGs reseeded.  That usually means there are several and I'm just not thinking straight enough.

paulidale commented:
Another possibility would be changing the CRYPTO_THREAD_xxx_lock calls to accept a NULL argument, returning 1 ?  I'm not sure I like this.  Nothing else is coming to mind at the moment.

The cost of using an unlocked lock is implementation and architecture dependent.  The worst case would likely be a context switch to the operating system (which is extremely expensive), the best an atomic memory operation (which is a lot more than a normal instruction but not thousands of cycles).

richsalz commented:
i think the master forcing a reseed on the two global ones, and using the regular locking, is probaby the good-enough trade-off for this work.  I really really like it.  I will remove more closely when i get home.  I will not be upset if it gets approved and merged before then :)

richsalz commented:
BTW, the writeup you put into your PR's is awesome.  I think the commit message might want to explicitly point to the PR text for more details.

mspncp commented:
Rich, your compliments make me blush again... :blush:

mspncp commented:
Ups, the cla-check is failing again?!

mspncp commented:
But seriously, I believe that an important aspect of a pull request is to convince other people and make them agree with the change. For a trivial pull request, the patch will be self explanatory, but for a larger work-in-progress, it is probably always a good idea to explain what one is doing.

dot-asm commented:
I'm not going do remove cla-check, nor do I want others to do so. I'd like to see/verify that next commit will remove it automatically...

mspncp commented:
Pushed and removed an additional commit to trigger the cla-check

mspncp commented:
## OS reseeding interval

One question to everybody: With the private and public DRBG now being chained, the number of generate requests (`RAND_bytes()` resp. `RAND_priv_bytes()`) between two OS reseeds now effectively jumps from 2^24 to 2^48. Does everybody agree that one should compensate this change by reducing the reseed_interval of the three global DRBG's (and only these) accordingly?

Please vote:

1. Leave all reseeding intervals at 2^24.
2. Change the reseeding intervals of the three global DRBG's to 2^12.
3. Change the reseeding interval of the public and private DRBG to 2^16, and of the master DRBG to 2^8.
4. Other choice.

richsalz commented:
if we make them all 2^8, then reseeding happens every 2^16 private keygens.
I think boring reseeds before every private keygen but I think that's too much
I mean, the "official math" says for the mode we use 2^24, right?

paulidale commented:
I suspect something that uses both a time based reseed and a bytes produced reseed.  I.e. **if** time since last reseed > X **or** number of bytes > Y **then** reseed.  The values for X and Y should perhaps be via a setter/getter interface.  2^48 for Y seems way too much.


For long life private stuff, an explicit reseed is beneficial.  I've got an item on my to do list to create some fast low quality non-blocking sources that can be added in this situation to provide some differentiation even if it isn't cryptographically good.  The idea being to mutate the state to make things more difficult.  I'm thinking things like gettimeofday, clock_gettime, TSC, RDRAND, /dev/urandom for this.  A full reseed every time is better from a security viewpoint but isn't practical.

If the caller requests prediction resistance, then a proper reseed from the OS should always be done.  I think this is mentioned in NIST's SP 800-90A.

Both these will be new PRs.

mspncp commented:
Just in case anybody wonders why this PR is stalled: The reviewers are keeping my busy on #4328 ;-)

paulidale commented:
I misunderstood slightly about the 2^48 in my previous response.

NIST's claim is that DRBGs are secure for 2^24 bytes.  I don't see a reason to dispute this, so the status quo is likely fine.  If we want to reduce this, we can but I'd not go too low due to the overhead involved in reseeding.  2^16 should be fine.  2^8 seems a bit too frequent.

NIST permits chains of DRBGs without crediting all of the output from all the DRBGs to the root of the chain.  For example, a chain of five DRBGs would reseed the root every 2^120 bytes output but at any point no DRBG will have produced more than 2^24 which is safe.

I still suggest a time based reseed lifetime as well, but again it doesn't have to be extremely frequent.

mspncp commented:
The code does not compile yet and still needs testing. I commited nevertheless to give you a short update about where I'm heading to in this PR.

In addition to automatic reseeding based on generate counts and seed propagation from master to its children, I also implemented an automatic reseeding of the master DRBG based on elapsed time.

See the commit message for more details. (And yes: the contents of the commit message should better go into a manual page...)

mspncp commented:
I chose to lower the reseeding interval of the public and private DRBG to 2^16, and of the master DRBG to 2^8. This ensures that randomness from the os entropy source is queried every 2^24 generate requests.

From the NIST point of view this would not be necessary, since a DRBG with access to a trusted entropy source is more or less equivalent to a NDRBG. So  it would be sufficient to reseed from the OS every 2^48 requests. However, in practice there is an essential difference: all three DRBGs live in the address space of the user program, whereas the OS RNG lives in the kernel, so its internal state is more secured than the state of the user space DRBGs: If one of the DRBG states gets leaked, the states of the other two  might not be secret anymore either.

kroeckx commented:
Can you rebase this now that the other commits have been merged?

mspncp commented:
Of course, this PR will be my next focus. Just give me a day.

mspncp commented:
Rebased onto the master branch. The implementation itself should be complete and functional.

mspncp commented on crypto/rand/drbg_lib.c:
```diff
@@ -33,9 +28,79 @@ static const char ossl_pers_string[] = "OpenSSL NIST SP 800-90A DRBG";
+ *    RAND_add() ==>    <master>          \
+ *                       /   \            | shared DRBGs
+ *                      v     v           |  (with locking)
+ *                 <public>  <private>    /
```
In the long term, this documentation should be moved to a proper place in the manual pages. How about adding a drbg.pod to man7 explaining all the general concepts, in addition to adding the RAND_DRBG API documentation in man3? (This will be done in a separate PR which publishes the API).

mspncp replied:
>  How about adding a drbg.pod to man7 explaining all the general concepts

This manual page would not only contain this comment, but a general overview over the entire new DRBG API, with some general concepts explained and references to the NIST document.

richsalz replied:
Great.  Please make all special-formatted comments start with "/*+" so that a future indent run, if/when it happens, doesn't mangle the picture and such.

mspncp commented on crypto/rand/rand_lcl.h:
```diff
@@ -20,8 +20,16 @@
-# define MAX_RESEED                     (1 << 24)
+/* Maximum reseed intervals */
+# define MAX_RESEED_INTERVAL                     (1 << 24)
+# define MAX_RESEED_TIME_INTERVAL                (1 << 20) /* approx. 12 days */
```
This is a somewhat arbitrary choice. On one hand, I did not want to admit ridiculously large reseed time intervals, on the other hand I did not want to be too strict.

mspncp commented on crypto/rand/drbg_lib.c:
```diff
@@ -164,7 +229,14 @@ int RAND_DRBG_instantiate(RAND_DRBG *drbg,
+        if (drbg->parent == NULL)
+            drbg->reseed_time = time(0);
+        else
+            drbg->reseed_time = drbg->parent->reseed_time;
```
It is clear to me that the child DRBG will not reseed more than once per second, even if the parent does, e.g., because RAND_add() is called more than once per second. But I believe that does not matter and I wanted to keep the implementation simple.

richsalz replied:
NULL not 0.  And maybe ```time(&drbg->reseed_time)```

mspncp replied:
Sometimes I just can't hide the fact, that I'm more used to C++. I will stick to your first suggestion, because as a mathematician I prefer the assignment. ;-)

mspncp replied:
And, yes, I kow 0 is out, everybody uses nullptr now ;-)

mspncp commented on crypto/rand/drbg_lib.c:
```diff
@@ -636,7 +759,7 @@ static int drbg_seed(const void *buf, int num)
 {
     int ret;
-    RAND_DRBG *drbg = RAND_DRBG_get0_global();
+    RAND_DRBG *drbg = RAND_DRBG_get0_public();
```
This should be RAND_DRBG_get0_master() here , for symmetry with drbg_add().

mspncp commented on crypto/rand/drbg_lib.c:
```diff
@@ -239,7 +311,7 @@ int RAND_DRBG_reseed(RAND_DRBG *drbg,
         goto end;
     drbg->state = DRBG_READY;
-    drbg->reseed_counter = 1;
+    drbg->reseed_counter = 0;
```
Just in case anybody wonders: This fixes an off-by-one error introduced by me when removing the unorthodox exit condition [at the end of RAND_DRBG_generate()](https://github.com/openssl/openssl/commit/e0b625f9db00509af9004b7907d44b78f332754a#diff-9181ac017a6177a5f2619f65c9b7a346) (line 405/407) in commit e0b625f9db00509af9004b7907d44b78f332754a.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -33,9 +28,79 @@ static const char ossl_pers_string[] = "OpenSSL NIST SP 800-90A DRBG";
+/*
+ * The public DRBG
+ *
+ * Used by default for generating random bytes using RAND_bytes().
```
Maybe you should add a note somewhere that we use it internally for numbers that become public, like in the RAND_bytes man page, plus expand on the history.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -33,9 +28,79 @@ static const char ossl_pers_string[] = "OpenSSL NIST SP 800-90A DRBG";
+ *
+ * This implies that a reseeding of the master DRBG is automatically
+ * propagated to the two other shared DRBGs, but not to the thread or
+ * SSL specific DRBGs.
```
So I think that means if someone using the private DRBG requests prediction resistance, all 3 would get reseeded? I would expect the private and the master to get reseeded (based on my understanding of the NIST documents), but from what I understand here the next time the public one is used it would also get reseeded. It's probably not a side effect we want.

mspncp replied:
No, currently the prediction_resistance request is not propagated to the parent. So a generate request to the private DRBG with prediction_resistance=1 only reseeds the private DRBG, not the master DRBG (unless its reseed_count reaches the reseed_interval).


Up to now it was my understanding that the NIST standard does not mandate the prediction request to be propagated to the parent DRBG: The NIST standard only considers a single DRBG. An adversary might have gained knowledge about its internal state, but not about its parent, which is considered a black box. Here are some citations from [NIST SP 800-90Ar1](http://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-90Ar1.pdf):

*Definition of Prediction Resistance (page 6)*

> An RBG provides prediction resistance relative to time T if it provides assurance that an adversary with knowledge of the state of the RBG at some time(s) prior to T (but incapable of performing work that matches the claimed security strength of the RBG) would be unable to distinguish between observations of ideal random bitstrings and (previously unseen) bitstrings output by the RBG at or subsequent to time T. In particular, an RBG whose design allows the adversary to step forward from the initially compromised RBG state(s) to obtain knowledge of subsequent RBG states and the corresponding outputs (including the RBG state and output at time T) would not provide prediction resistance relative to time T.

On page 25, it says 

> The prediction_resistance_request parameter indicates whether or not prediction resistance is to be provided during the request (i.e., whether fresh entropy is required[4]).

Chaining of DRBGs is justified by footnote[4] of page 25:

> [4] 4 Entropy input may be obtained from an entropy source or an NRBG, both of which provide fresh entropy. Entropy
input could also be obtained from a DRBG that has access to an entropy source or NRBG. The request for prediction resistance rules out the use of a DRBG that does not have access to either an entropy source or NRBG.

and in further detail by footnote [5], page 30:
> A DRBG may be reseeded by an entropy source or an NRBG, both of which provide fresh entropy. A DRBG could also be reseeded by a DRBG that has access to an entropy source or NRBG. The request for prediction resistance during reseeding rules out the use of a DRBG that does not have access to either an entropy source or NRBG. See [SP 800-90C] for further discussion. 

In our case one might argue differently, because the master DRBG resides in the same address space as public and private DRBG. An adversary that gains knowledge about the private DRBG could indeed also have knowledge about its parent. I tried to mitigate this possility by lowering the reseed intervals to keep the total "os reseed interval" at 2^24. But I would understand if someone insisted on forwarding the prediction_resistance request to the parent. In this case one would indeed get the Domino effect you mentioned. It's the usual security vs. performance tradeoff. As such, it's a political question and should probably be decided the OMC members. I am a little in favor of my choice, but would also follow an majority vote.

kroeckx replied:
I have no idea how you can interpret that text differently. There needs to be a way to get fresh entropy from somewhere. A DRBG does not provide fresh entropy, you need something that directly talks to the entropy source to get such fresh entropy and that's the master DRBG.

paulidale replied:
My understanding of this standard was that the request for new entropy should propagate along the chain until some fresh entropy had been gathered from a non-DRBG source.

I.e. the private DRBG should cause the master DRBG to reseed before reseeding itself.

mspncp replied:
I disagree, and it is not a matter of interpretation, it is written down in the standard:

- Requesting prediction resistance simply means requesting a reseeding of the DRBG before generating random output. There is no difference between a "normal" reseeding and a reseeding triggered by a prediction resistance request.
- The NIST standard explicitely allows reseeding from a DRBG which is connected to an approved entropy source. There is no word about a need to reseed the parent immediately before obtaining its random output.

I you are questioning the fact that a DRBG can safely be reseeded from the master DRBG, then you question the entire construction of chaining the DRBGs to improve the security of the DRBGs in the first place.

But before going deeper into this, I would like to dispel your concerns about prediction resistance:

I agree that a kernel mode CSPRNG is more secured than a user space CSPRNG. So I completely understand if someone prefers to pull randomness from the former instead of the latter. But this is already possible to do: If the applications needs "strong" prediction resistance for the public or the private DRBG, all it has to do is to call RAND_poll() before calling RAND_bytes() resp. RAND_priv_bytes(). This will reseed the master DRBG from the os entropy sources, and the reseeding will then be propagated to the public resp. private DRBG. Alternatively, it could obtain a pointer to the master DRBG (drbg = RAND_DRBG_get0_master()) and call its generate method requesting prediction resistance.


### The facts

I was not trying to interpret the standard, I was stating the facts from the document and verifying them by giving quotations. So please tell me, where exactly I misinterpreted something in my reasoning below. And please point out the reference where it is stated that the parent DRBG's generate method SHALL be called demanding prediction resistance.

The DRBG construction is an approved method of 'stretching' the sparse resource of randomness into a longer pseudo random output. This means that the DRBG's output is computationally indistinguishable from a true random coin toss for an adversary whose computational capabilities lie within the limits assumed for the given security level (not more and not less). This holds under the assumptions that

- the DRBG was reseeded from an approved entropy source or nondeterministic RNG
- the internal state of the DRBG is unknown to the adversary

If the previous statement holds then not only the first block of 128 bits generated by the parent DRBG is computationally indistinguishable from true random output, but also the second and the third block, etc. So why should the first 128 bit block contain "more entropy" than the second? Neither of them contains any entropy because the output is a deterministic function of the internal state. All the randomness lies in the random choice of this internal state, and the fact that the internal state remains secret.

Nothing else happens when you reseed from the kernel's RNG: You will get some block of pseudo random output from a deterministic RNG and you can't even be certain that enough bits of "fresh" entropy have been added to the kernel's random pool since your last request (more on this subject below).

Calling RAND_DRBG_generate() with prediction_resistance=1 is the request to "reseed before generating output", in order to handle the case that the internal state of the DRBG might have been (partially) compromised. It triggers a "normal" reseed, without any additional constraints on the "quality" of the random input pulled from the entropy source.

The NIST standard explicitely states that instead of an approved nondeterministic RNG or entropy source it is allowed to use a DRBG which is seeded by an approved nondeterministic RNG or entropy source. Indeed, unless the state of the parent DRBG is compromised, too, or the randomness input fed into the child DRBG is intercepted by the adversary, this reseeding will immediately restore the integrity of the child DRBG.


### The myth

What is "fresh" entropy in the first place? Is it filled into the DRBG and used up by generate requests like the gasoline of an electricity generator (pun intended)? No, this is just and analogy for humans and taking this analogy too literal is the cause for the misbelief that a DRBG which has been seeded with, say, 128 bits of entropy gets drained after generating 128 bits of random output. This superstition caused generations of programmers to block on /dev/random [waiting for "fresh" entropy](https://www.2uo.de/myths-about-urandom/#low-entropy). I don't know the author of the linked article, but I agree with his arguments, in particular with the [quote of Daniel Bernstein](https://www.2uo.de/myths-about-urandom/#experts), which was taken from [this email](https://www.mail-archive.com/cryptography@randombit.net/msg04763.html)).


### Theory vs. practice

In the context of the NIST document the case is clear: it is not necessary to propagate the prediction resistance request to the master. In our case the situation is a little more complicated, because both DRBGs are located in the address space of the same process, so it is not clear whether the master DRBG's internal state can be considered secure after the child DRBG's internal state was compromised.

Whether or not the master DRBG can be considered secure depends on the threat model. If the adversary can only intercept network traffic and learn about the state of the child DRBG, but does not get to see any of the master DRBG's output, then the answer is clearly yes. (Apropos: it's probably better not to reseed the ssl DRBGs from the public DRBG, because we have no control how much output of the public DRBG will be used for other purposes and become publicly known.)

If the adversary obtained knowledge of the child DRBG's internal state by reading some portion of the process memory, then it is arguable that he might now about the internal state of the master, too. One mitigation for this threat would be to not use static variables to hold the internal state of the three DRBGs, but allocate it from the secure heap.

Anyway, as stated above: It is easy to trigger a full reseed from the os entropy sources either by calling RAND_poll() or by using a DRBG which is not chained and requesting prediction resistance.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -164,7 +229,14 @@ int RAND_DRBG_instantiate(RAND_DRBG *drbg,
-    drbg->reseed_counter = 1;
+    drbg->reseed_counter = 0;
+
+    if (drbg->reseed_time != 0) {
```
Why the if?

kroeckx replied:
It seems that when it's set to 0 it's not used. Please document that somewhere.

kroeckx replied:
I guess I would prefer that reseed_time_interval is used for that instead.

kroeckx replied:
I don't see why it's only meaningful for a non-chained DRBG. You set the interval in both cases, and when it's enabled it's always different from 0. It's just that in a chained DRBG you don't actually look at the interval value itself.

Anyway, it's fine like this.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -315,7 +387,7 @@ int rand_drbg_restart(RAND_DRBG *drbg,
     /* repair uninitialized state */
     if (drbg->state == DRBG_UNINITIALISED) {
-        drbg_setup(drbg, NULL);
+        drbg_setup(drbg, NULL, NULL);
```
Should that be drbg->parent?

kroeckx replied:
Oh, it's using code that doesn't actually update the parent when called with NULL ...

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -397,6 +469,19 @@ int RAND_DRBG_generate(RAND_DRBG *drbg, unsigned char *out, size_t outlen,
+            if (drbg->reseed_time != drbg->parent->reseed_time)
+                reseed_required = 1;
+        } else {
+            time_t now = time(0);
```
time(NULL)

kroeckx commented on crypto/rand/rand_lib.c:
```diff
@@ -247,7 +247,7 @@ int RAND_poll(void)
     if (meth == RAND_OpenSSL()) {
         /* fill random pool and seed the default DRBG */
-        RAND_DRBG *drbg = RAND_DRBG_get0_global();
+        RAND_DRBG *drbg = RAND_DRBG_get0_public();
```
This should probably be master.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -194,7 +267,6 @@ int RAND_DRBG_uninstantiate(RAND_DRBG *drbg)
 {
     int ret = ctr_uninstantiate(drbg);
 
-    OPENSSL_cleanse(&drbg->ctr, sizeof(drbg->ctr));
```
Is there a reason for this?

mspncp replied:
It was moved into ctr_uninstantiate(), where it replaces a memset. To understand the change, its best to have a view at changeset https://github.com/openssl/openssl/pull/4402/commits/aec27cfcc5077171cc09344fe6d4aa6eab7479be

kroeckx commented on ssl/ssl_lib.c:
```diff
@@ -635,7 +635,7 @@ SSL *SSL_new(SSL_CTX *ctx)
         s->drbg =
             RAND_DRBG_new(RAND_DRBG_NID, RAND_DRBG_FLAG_CTR_USE_DF,
-                          RAND_DRBG_get0_global());
+                          RAND_DRBG_get0_public());
```
Should we enable the auto reseeding for the ssl specific DRBG?

mspncp replied:
It was @paulidale's whish not to do it because he feared that a reseed of the master DRBG would cause a chain reaction. See https://github.com/openssl/openssl/pull/4402#discussion_r140406758 and my reply https://github.com/openssl/openssl/pull/4402#discussion_r140435339.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -472,16 +558,51 @@ int RAND_DRBG_set_callbacks(RAND_DRBG *drbg,
+ * For a child DRBG, the |interval| argument is just used to enable or disable
+ * the reseed timer, depending on whether its value is nonzero or not.
+ * If the reseed timer is enabled, the DRBG will reseed itself whenever its
+ * reseed_time differs from the parent's reseed_time.
```
I guess that "reseed timer" is confusing, since there actually is no timer involved.
It's a way for a child DRBG to reseed when the parent DRBG reseeded.
I'm wondering if we should split the functionality to make it more clear what it does.

mspncp replied:
You are right. I wanted to replace "reseed timer" by "automatic reseeding", but forgot to do it:

```
 * For a child DRBG, the |interval| argument is just used to enable or disable
 * automatic reseeding, depending on whether its value is nonzero or not.
 * If automatic reseeding is enabled, the DRBG will reseed itself whenever its
 * reseed_time differs from the parent's reseed_time.
```

> I'm wondering if we should split the functionality to make it more clear what it does.

I was pondering about that variant, too. But I abandoned it, because the current implementation appeared simpler to me. I think the implementation itself is ok, only the explanation is not clear enough. So I will try to improve the comments.

mspncp commented:
Rebased the branch and reorganized the fixups a little bit to group the hunks logically. Only the last two fixups contain new stuff.

Currently trying to figure out how to implement the tests for seed chaining...

mspncp commented on crypto/rand/drbg_lib.c:
```diff
@@ -382,12 +382,20 @@ int rand_drbg_restart(RAND_DRBG *drbg,
 
     /* repair uninitialized state */
     if (drbg->state == DRBG_UNINITIALISED) {
-        drbg_setup(drbg, NULL, NULL);
```
This was a hack, see commit message for https://github.com/openssl/openssl/pull/4402/commits/ff9e120e0d38d9e10ee498ad2d01bafca38bd55f - rand_drbg_restart: reinstantiate drbg without using drbg_setup() [fixup]

mspncp commented on crypto/rand/drbg_lib.c:
```diff
@@ -33,9 +28,79 @@ static const char ossl_pers_string[] = "OpenSSL NIST SP 800-90A DRBG";
+ *
+ * There are three shared DRBGs (master, public and private), which are
+ * accessed concurrently by all threads.
+ *
```
I was wondering, whether the three shared DRBGs shouldn't better be allocated on the secure heap (CRYPTO_secure_*()) instead of having them here as static variables, probably even placed in consecutive memory by the compiler.

paulidale replied:
This makes sense.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -310,12 +382,20 @@ int rand_drbg_restart(RAND_DRBG *drbg,
     if (drbg->state == DRBG_UNINITIALISED) {
-        drbg_setup(drbg, NULL);
+        /* reset reseed time */
+        drbg->reseed_time = 1;
```
Does that need an `if (drbg->reseed_time != 0)`?

mspncp replied:
Yes, indeed!

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -501,31 +630,42 @@ void *RAND_DRBG_get_ex_data(const RAND_DRBG *drbg, int idx)
 /*
- * Initializes the DRBG with default settings.
+ * Initializes the given global DRBG with default settings.
  * For global DRBGs a global lock is created with the given name
```
I think it's now only used for global DRBGs and the lock is always created, so this comment seems out of date.

mspncp replied:
Canged to 

```
 * Initializes the given global DRBG with default settings.
 * A global lock for the DRBG is created with the given name.
```

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -310,12 +382,20 @@ int rand_drbg_restart(RAND_DRBG *drbg,
+    if (drbg->state == DRBG_ERROR) {
         RAND_DRBG_uninstantiate(drbg);
+        /* The drbg->ctr member needs to be reinitialized before reinstantiation */
+        RAND_DRBG_set(drbg, drbg->nid, drbg->flags);
```
I assume the reason for this is so that ctr_init() gets called again. Maybe we can call that in RAND_DRBG_instantiate()? Or move this call to the next if?

kroeckx replied:
Any comment about this?

mspncp commented:
### Add Tests for the automatic reseeding
    
A new test `test_rand_reseed` was added which checks some standard cases where seed propagation should occur.

- Test whether all three DRBGs are reseeded after reseed_time_interval
- Test whether the public resp. private DRBG is reseeded when its reseed_counter differs from the master's reseed_counter.
- Test whether all three DRBGs are reseeded by RAND_add()


### Enhance automatic reseeding
    
During the development of the tests two issues showed up:
    
 - Even if the reseed_time_interval of the master has been reached, the public and private DRBG would not immediately be reseeded when calling RAND_bytes() and RAND_priv_bytes(), respectively. This is because the child DRBGs did not check the system time, but only compared their reseed_time to the master's reseed_time. So they would continue generating random bytes without reseeding until the reseed_interval had been reached.
    
 - The granularity of the reseed_time is just in seconds, so when the sequence RAND_add(); RAND_bytes() occurs multiple times whithin a second, ony the first RAND_add() call will be forwarded to the public DRBG and change the output of RAND_add(). Although this is really security issue, because the RAND_add() calls get accumulated until the next second when the child DRBG reseeds, it is nevertheless confusing, so it was better to fix this.
    
To solve these two issues, the following change was made:

In addition to the reseed_time, a new reseed_counter was introduced,
which is incremented by the master on every reseed. The child now also
checks the system time and reseeds in any of the two following cases:

 - The reseed_counter differs from its parent's reseed_counter
 - The number of seconds since reseed_time has reached reseed_time_interval.

After every reseed, the child copies both the reseed_time and the
reseed_counter from its parent.

mspncp commented:
### Rename reseed_counter -> generate_counter

To avoid a name collision with the reseed_counter introduced by the commit "Enhance automatic reseeding", the existing `drbg->reseed_counter` was renamed to `drbg->generate_counter`. This name is much more appropriate, since the counter is used to count the number of generate requests since the last reseed. The only thing that made me hesitate renaming the counter is that its name occurs in the NIST standard. However, keeping the name would have forced me to choose a stupid alternative name for the new reseed counter.

If anyone objects to this renaming, please let me now.  (And please provide a reasonable name for the new counter ;-) )

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -164,7 +224,18 @@ int RAND_DRBG_instantiate(RAND_DRBG *drbg,
+            drbg->reseed_time = time(NULL);
+            drbg->reseed_counter = 0;
+        } else {
+            drbg->reseed_time = drbg->parent->reseed_time;
```
Does the child still need to track the parent's reseed_time now that we have a reseed_counter?

kroeckx replied:
So I guess the answer is yes, so next comment.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -501,31 +647,43 @@ void *RAND_DRBG_get_ex_data(const RAND_DRBG *drbg, int idx)
+    } else {
+        drbg->parent = parent;
+        drbg->reseed_interval = SLAVE_RESEED_INTERVAL;
+        drbg->reseed_time_interval = parent->reseed_time_interval;
```
This is to ensure that the we actually call the parent DRBG when his time is up? This is non-obvious and could use some comment.

mspncp replied:
Yes this precisely the point. It fixes the first issue mentioned in the commit message of b1979bf77d718d16e04b30b00ef96c9193d9c2f4. But since that commit and its message will be squashed, it's better to have some explaining comments here and in rand_lcl.h, as you suggested. I will add them soon.

kroeckx commented on crypto/rand/rand_lcl.h:
```diff
@@ -106,8 +114,24 @@ struct rand_drbg_st {
+    /* maximum number of generate requests until a reseed is required */
     unsigned int reseed_interval;
+    /* counts the number of reseeds since instantiation */
+    unsigned int reseed_counter;
```
Can you add that that is for the parent and that for the child it's the parent's reseed counter?

mspncp replied:
Yes, I will add a more precise explanation.

mspncp commented:
Rebased to resolve conflicts. I squashed the older fixups, because fixing multiple libcrypto.num conflicts became really painful.

mspncp commented:
Ready for final review.

mspncp commented:
Ping @openssl

mspncp commented:
@kroeckx, what's the current state of your review?

kroeckx commented:
I'll try to look at it tomorrow.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -394,9 +489,23 @@ int RAND_DRBG_generate(RAND_DRBG *drbg, unsigned char *out, size_t outlen,
+    if (drbg->reseed_time > 0) {
+        /* automatic reseeding is enabled */
+        if (drbg->parent != NULL
+            && drbg->reseed_counter != drbg->parent->reseed_counter)
```
Shouldn't this also check the parent->reseed_time?

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -394,9 +489,23 @@ int RAND_DRBG_generate(RAND_DRBG *drbg, unsigned char *out, size_t outlen,
+            if (now < drbg->reseed_time
+                || now - drbg->reseed_time >= drbg->reseed_time_interval)
+                reseed_required = 1;
+        }
```
I think the whole block should not be in the else, both the parent and child should check the reseed_time_interval.

kroeckx replied:
At least based on other comment in the file, that's not really what we want. We want to check that the master's reseed_time_interval is up at some place, and then make sure a generate call is made. This doesn't seem to be happening anywhere currently. One way of doing that is check the parent here and if needed set reseed_required.

kroeckx replied:
Oh, but our reseed_time equals the parent's reseed_time, so removing the else is what we wanted anyway.

kroeckx replied:
Which makes me wonder why we copy this, while we could just look at it? And that really doesn't seem to work when there are 3 DRBGs in the chain?

kroeckx replied:
I guess I see 2 options:
- Actually look at the parent's variables. This will probably require taking a lock which we want to avoid here.
- Split the variables, so there is a set for the current drbg and the parent drbg

kroeckx replied:
Having a copy doesn't work for the RAND_add() case

mspncp replied:
> And that really doesn't seem to work when there are 3 DRBGs in the chain?

You are right, this does not work, but it was never intended to. See my review summary.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -544,34 +729,42 @@ static int drbg_setup(RAND_DRBG *drbg, const char *name)
-    int ret = 1;
+    drbg_master  = drbg_setup("drbg_master",  NULL);
+    drbg_public  = drbg_setup("drbg_public",  drbg_master);
+    drbg_private = drbg_setup("drbg_private", drbg_master);
```
I'm not really a fan of the extra spaces

mspncp replied:
Added a fixup to remove the extra spaces.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -238,8 +312,20 @@ int RAND_DRBG_reseed(RAND_DRBG *drbg,
+            drbg->reseed_counter++;
+        } else {
+            drbg->reseed_time = drbg->parent->reseed_time;
+            drbg->reseed_counter = drbg->parent->reseed_counter;
```
I think we should take drbg->parent's lock before accessing any of it's variables both here and at other places.

mspncp replied:
There was a lengthy discussion with @paulidale and @dot-asm in PR #4414 whith that result hat reading the reseed count does not have to be atomic, see https://github.com/openssl/openssl/pull/4414#issuecomment-332025551 and https://github.com/openssl/openssl/pull/4414#issuecomment-332460299. For the same reasons, I don't see any benefit from taking a lock here.

kroeckx commented on crypto/rand/rand_lcl.h:
```diff
@@ -106,8 +114,36 @@ struct rand_drbg_st {
+     * For details, see the comment before RAND_DRBG_enable_auto_reseed().
+     */
+
+    /* Counts the number of seconds since the last reseed */
```
Stores the time of the last reseed?

mspncp replied:
Ups, I missed this one. A stupid copy and paste error. Will be fixed in a moment.

mspncp commented:
As announced in my previous post, I reworked and cleaned up this PR. In particular, I untangled the automatic reseeding of a DRBG based on the elapsed time (`reseed_time_interval`) and the seed propagation from master DRBG to public and private DRBG  (`reseed_counter`). They are now implemented in separate commits.

Hope this makes the PR cleaner and more comprehensible. The four commits are not intended to be squashed.

@paulidale, please reconfirm.

mspncp commented:
Travis is reporting a lot of seemingly unrelated errors:

```
../../util/shlib_wrap.sh ../../apps/openssl ecparam -noout -check -in ../../../_srcdist/test/recipes/15-test_ecparam_data/valid/wap-wsg-idm-ecid-wtls9-named.pem => 0
ok 130
unable to load elliptic curve parameters
47969742222976:error:10107010:elliptic curve routines:EC_GROUP_new_from_ecparameters:EC lib:../_srcdist/crypto/ec/ec_asn1.c:787:
47969742222976:error:10108010:elliptic curve routines:EC_GROUP_new_from_ecpkparameters:EC lib:../_srcdist/crypto/ec/ec_asn1.c:856:
47969742222976:error:1009107F:elliptic curve routines:d2i_ECPKParameters:pkparameters2group failure:../_srcdist/crypto/ec/ec_asn1.c:885:
47969742222976:error:0906700D:PEM routines:PEM_ASN1_read_bio:ASN1 lib:../_srcdist/crypto/pem/pem_oth.c:33:
```

mspncp commented:
The [windows x86 build](https://ci.appveyor.com/project/openssl/openssl/build/master.14198/job/96834felprtxuxao) fails because I am using `TEST_int_le()` to compare `time_t` values:

```
    if (!TEST_int_le(before_reseed, master->reseed_time)
        || !TEST_int_le(master->reseed_time, after_reseed))
        return 0;
```
and the build is treating warnings as errors:

```
[00:03:50] 	cl  /I "include" /I "test" /I "." /I "..\include" /I "..\test" /I ".." -DOPENSSL_USE_APPLINK -DDSO_WIN32 -DNDEBUG -DOPENSSL_THREADS -DOPENSSL_NO_STATIC_ENGINE -DOPENSSL_PIC "-DENGINESDIR=\"C:\\Program Files (x86)\\OpenSSL\\lib\\engines-1_1\"" "-DOPENSSLDIR=\"C:\\Program Files (x86)\\Common Files\\SSL\"" -W3 -wd4090 -Gs0 -GF -Gy -nologo -DOPENSSL_SYS_WIN32 -DWIN32_LEAN_AND_MEAN -DL_ENDIAN -D_CRT_SECURE_NO_DEPRECATE -D_WINSOCK_DEPRECATED_NO_WARNINGS -DUNICODE -D_UNICODE /MD /O2 -WX /Zi /Fdapp -c /Fotest\drbgtest.obj "..\test\drbgtest.c"
[00:03:50] drbgtest.c
[00:03:51] ..\test\drbgtest.c(601): error C2220: warning treated as error - no 'object' file generated
[00:03:51] ..\test\drbgtest.c(601): warning C4244: 'function': conversion from 'time_t' to 'const int', possible loss of data
[00:03:51] ..\test\drbgtest.c(602): warning C4244: 'function': conversion from 'time_t' to 'const int', possible loss of data
```

Unfortunately, there seems to be no comparison function for `time_t`. And it's probably not so easy to implement one correctly for all platorms (see `size_t`). @levitte, @paulidale, do you have a good suggestion for a workaround or a some portable time_t comparison operators for me?

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -64,7 +129,7 @@ int RAND_DRBG_set(RAND_DRBG *drbg, int nid, unsigned int flags)
 
-    if (ret < 0)
+    if (ret == 0)
         RANDerr(RAND_F_RAND_DRBG_SET, RAND_R_ERROR_INITIALISING_DRBG);
```
crt_init can return -1, not 0.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -164,7 +229,14 @@ int RAND_DRBG_instantiate(RAND_DRBG *drbg,
-    drbg->reseed_counter = 1;
+    drbg->generate_counter = 0;
+    drbg->reseed_time = time(NULL);
+    if (drbg->reseed_counter > 0) {
```
Why this if?

kroeckx replied:
I can not find any place where this gets set value different than 0.

mspncp replied:
You are completely right, good catch! This got lost somehow  in my last overhaul when I untangled time-based reseeding and seed propagation from master to public and private DRBG. The `drbg->reseed_time` used to be non-zero, but the `drbg->reseed_counter` isn't anymore. It needs to be set explicitely to 1 in `drbg_setup()` in order to enable seed propagation for the shared DRBGs.

What worries me a bit that this glitch was not noticed by my tests which explicitely check whether seed propagation works. I'll have to look into this.

kroeckx replied:
The problem in the test is of course that previous tests set those counters to something other than 0.

mspncp replied:
Yes, I noticed. Meanwhile the test has been modified such that the reseed_counters are not modified anymore before the initial call to test_drbg_reseed() (see drbgtest.c:693).

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -164,7 +229,14 @@ int RAND_DRBG_instantiate(RAND_DRBG *drbg,
+    drbg->reseed_time = time(NULL);
+    if (drbg->reseed_counter > 0) {
+        if (drbg->parent == NULL)
+            drbg->reseed_counter = 1;
```
I'm wondering if it wouldn't be better to do ++ here

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -238,8 +315,16 @@ int RAND_DRBG_reseed(RAND_DRBG *drbg,
+        if (drbg->parent == NULL)
+            drbg->reseed_counter++;
+        else
+            drbg->reseed_counter = drbg->parent->reseed_counter;
```
Do we intent to support chains with more than 2 DRBGs where we expect each child to reseed once the parent reseeds?

This could probably also use a line of comment.

mspncp replied:
This is correct. But I think the revised comments in rand_lcl.h should now be sufficient, so this does not have to be restated here again. Moreover since any such comment would have to be repeated in RAND_DRBG_instantiate(), which contains similar code.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -394,8 +482,20 @@ int RAND_DRBG_generate(RAND_DRBG *drbg, unsigned char *out, size_t outlen,
+    if (drbg->reseed_counter > 0 && drbg->parent != NULL) {
+        if (drbg->reseed_counter != drbg->parent->reseed_counter)
+            reseed_required = 1;
+    }
```
I think it should also check that the parent's reseed time is up

kroeckx replied:
If we intend to support more than 2 levels, that would actually mean checking all parents.

kroeckx replied:
As it is currently, they have the same reseed interval by default (1 hour), so they'll both get reseed at that time. Do we care that the master doesn't get reseeded when it's time is up and only reseeds the next time that we actually use it? I guess not. So it's probably fine like it is.

mspncp replied:
The design philosophy of the NIST DRBG standard is that every DRBG has its own reseeding strategy which does not depend on its parent's reseeding strategy. The parent DRBG is only used as entropy source, but there is no coupling between reseeding events.

The `reseed_counter` is a deviation from this principle. This is a compromise which was necessary to conciliate the different philosophies of the RAND_METHOD and the RAND_DRBG api. 

For the `reseed_interval` and the `reseed_time_interval` I don't intend to deviate from this simple principle. Since we have full control over reseed interval and reseed time interval of all three shared DRBGs, there is no need for the children to synchronize with the parent if we make reasonable choices for their default values.

The current defaults are just a first draft which maybe needs some finetuning.
```C
/* Default reseed intervals */
# define MASTER_RESEED_INTERVAL                  (1 << 8)
# define SLAVE_RESEED_INTERVAL                   (1 << 16)
# define MASTER_RESEED_TIME_INTERVAL             (60*60)   /* 1 hour */
# define SLAVE_RESEED_TIME_INTERVAL              (60*60)   /* 1 hour */
```
Maybe the slave reseed time interval should be much shorter and maybe it should not divide the master reseed interval, e.g.
```C
/* Default reseed intervals */
# define MASTER_RESEED_INTERVAL                  (1 << 8)
# define SLAVE_RESEED_INTERVAL                   (1 << 16)
# define MASTER_RESEED_TIME_INTERVAL             (60*60)   /* 1 hour */
# define SLAVE_RESEED_TIME_INTERVAL              (7*60)    /* 7 minutes */
```
In this matter I am open to and thankful for suggestions.

paulidale replied:
The latter feels nicer but I've no solid evidence either way.  It reseeds the slaves more often which shouldn't hurt.

I don't see a benefit from the coprimality but it isn't going to hurt either.

kroeckx commented on crypto/rand/drbg_lib.c:
```diff
@@ -501,31 +625,47 @@ void *RAND_DRBG_get_ex_data(const RAND_DRBG *drbg, int idx)
+ * initializes it with default settings.
+ * A global lock for the DRBG is created with the given name.
+ *
  * Returns 1 on success, 0 on failure
```
It returns a drbg on success and NULL on failure

mspncp replied:
Fixed, here and in the comment of RAND_DRBG_new().

paulidale commented:
Would a cast to _long_ work for the _time_t_ comparison?

I'll have a think about how to add a set of _time_t_ tests.  The problem with _size_t_ only looks half implemented.  There isn't a guard around the implementation, just the declaration which defeats trying to avoid the `%zu` format.

paulidale commented:
#4797 might help with the test.

mspncp commented:
> #4797 might help with the test.
Thanks a lot!

paulidale commented:
The time_t test functions are merged now.

mspncp commented:
Thanks a lot, I appreciate your help very much. I didn't have much time to continue working on this PR in the past few days, but I'm still on it.

mspncp commented:
ping @kroeckx and @paulidale: ready for final review.

paulidale commented on crypto/rand/drbg_lib.c:
```diff
@@ -501,31 +627,47 @@ void *RAND_DRBG_get_ex_data(const RAND_DRBG *drbg, int idx)
+    if (drbg == NULL)
+        return NULL;
+
+    drbg->lock = CRYPTO_THREAD_glock_new(name);
```
I wonder if this should be `CRYPTO_THREAD_lock_new`?  `CRYPTO_THREAD_glock_new` adds the lock to a list of locks that are unlocked on `fork(2)` so I think the answer is _no_.  However, there is currently no way to remove a lock from this list so the lock freeing later will cause problems accessing deallocated memory.

Don't try to fix this here.

paulidale commented on crypto/rand/rand_lcl.h:
```diff
@@ -20,8 +20,17 @@
+# define MASTER_RESEED_INTERVAL                  (1 << 8)
+# define SLAVE_RESEED_INTERVAL                   (1 << 16)
+# define MASTER_RESEED_TIME_INTERVAL             (60*60)   /* 1 hour */
+# define SLAVE_RESEED_TIME_INTERVAL              (60*60)   /* 1 hour */
```
Should we go with the seven minute suggestion?

mspncp replied:
Kurt, if you agree with the seven minute suggestion, then I will change the `SLAVE_RESEED_INTERVAL` to 7 minutes and squash all the fixup! commits after you finished reviewing.

kroeckx replied:
7 minutes would is fine for me.

paulidale commented:
I like it!

kroeckx commented:
I probably won't have time to look at this until at least Friday.

mspncp commented:
That's ok, no problem.

mspncp commented:
Added the 7 minute reseed time interval for public and private DRBG as fixup.

mspncp commented:
ping :sleeping:

mspncp commented:
I added checks for the reseed_count to the testsuite and verified that the testsuite indeed detects my error, before fixing it. Additionally, I added a check that simulates a failed reseeding of the master drbg: In this case, both RAND_bytes() and RAND_priv_bytes() calls should fail.

I rebased the branch to match the fixups properly to the commits. Autosquashing was not entirely automatic, some manual work was required. But I checked that every single commit of the rebased branch compiles cleanly and passes the drbgtest.

To keep things simple for you, I left the unsquashed branch in [mspncp:pr-add-master-drbg-unsquashed](https://github.com/mspncp/openssl/commits/pr-add-master-drbg-unsquashed). The unsquashed branch contains my final fixup (one commit) and you can verify that there are no differences between pr-add-master-drbg and pr-add-master-drbg-unsquashed.

mspncp commented:
Just added one more trivial fixup, see [mspncp:pr-add-master-drbg-unsquashed](https://github.com/mspncp/openssl/commits/pr-add-master-drbg-unsquashed):
```
commit 14092f45a35b64c395461c8ef152ee542df7da35 (HEAD -> pr-add-master-drbg-unsquashed, msp/pr-add-master-drbg-unsquashed)
Author: Dr. Matthias St. Pierre <Matthias.St.Pierre@ncp-e.com>
Date:   Wed Dec 13 15:43:02 2017 +0100

    Removed unneeded '#include <time.h>'  [fixup]

commit 44cdf3552784a136a58307d6bc2e49cbbfe8637c
Author: Dr. Matthias St. Pierre <Matthias.St.Pierre@ncp-e.com>
Date:   Wed Dec 13 12:53:38 2017 +0100

    fixup! Add master DRBG for reseeding
```

mspncp commented:
Fixup changes an erroneous `TEST_int_le()` call to `TEST_int_eq()`.

mspncp commented:
Kurt, do you think you‘ll find the time in the next days to complete your review? This PR will be open three months by the end of the week and it would be nice if we could have it finished before the holidays.

kroeckx commented:
I hope to complete it today.

kroeckx commented on test/drbgtest.c:
```diff
@@ -475,26 +475,295 @@ static int test_error_checks(int i)
+     */
+
+    /* Generate random output from the public and private DRBG */
+    before_reseed  = expect_master_reseed == 1 ? time(0) : 0;
```
It should be time(NULL), there is also an extra space.

(Can be fixed when merging)

kroeckx commented on test/drbgtest.c:
```diff
@@ -475,26 +475,295 @@ static int test_error_checks(int i)
+    if (!TEST_int_eq(RAND_bytes(buf, sizeof(buf)), expect_success)
+        || !TEST_int_eq(RAND_priv_bytes(buf, sizeof(buf)), expect_success))
+        return 0;
+    after_reseed = time(0);
```
time(NULL)

kroeckx commented on test/drbgtest.c:
```diff
@@ -475,26 +475,295 @@ static int test_error_checks(int i)
+
+    if (expect_private_reseed >= 0) {
+        /* Test whether public DRBG was reseeded as expected */
+        if (!TEST_int_eq(private_ctx.reseed_count,  expect_private_reseed))
```
extra space

kroeckx commented:
@paulidale: Can you confirm you're happy with this?

paulidale commented:
The recent changes look good.

I didn't go through everything again but can if desirable.

kroeckx commented:
I think we're still missing a test for the reseed_time_interval, but maybe we can do that in an other PR.

kroeckx commented:
@paulidale: It's really up to you to say if this is good to merge or not.

mspncp commented:
@kroeckx, @paulidale Thank you very much for reviewing my pr. 

I would like to add the final fixups myself, because they  belong to two different commits.  Please let me know if it's ok for you when I autosquash the fixups  (without moving the base) while leaving the unsquashed fix at [mspncp:pr-add-master-drbg-unsquashed](https://github.com/mspncp/openssl/commits/pr-add-master-drbg-unsquashed) for inspection. If you prefer to do the rebasing yourself, then please do it using  `git rebase -i --autosquash` after I added the fixups.

kroeckx commented:
Feel free to rebase/squash as needed

mspncp commented:
Autosquashed and rebased to the tip of master. Thanks again to everybody for taking the time to review this pr!

mspncp commented:
> I think we're still missing a test for the reseed_time_interval, but maybe we can do that in an other PR.

Maybe I'll find some time the next days to prepare a pr for the missing test.

paulidale commented:
Okay, I've gone through it again and still approve.
I'll merge.

paulidale commented:
Merged but _not_ squashed to a single commit.

paulidale commented:
Thanks for everyone's patience.
